src/strings.rs

```diff
@@ -0,0 +1,220 @@
+//! Types to support arrays of [ASCII][ascii] and [UCS4][ucs4] strings
+//!
+//! [ascii]: https://numpy.org/doc/stable/reference/c-api/dtype.html#c.NPY_STRING
+//! [ucs4]: https://numpy.org/doc/stable/reference/c-api/dtype.html#c.NPY_UNICODE
+
+use std::cell::RefCell;
+use std::collections::hash_map::Entry;
+use std::convert::TryInto;
+use std::fmt;
+use std::mem::size_of;
+use std::os::raw::c_char;
+use std::str;
+
+use pyo3::{
+    ffi::{Py_UCS1, Py_UCS4},
+    sync::GILProtected,
+    Py, Python,
+};
+use rustc_hash::FxHashMap;
+
+use crate::dtype::{Element, PyArrayDescr};
+use crate::npyffi::NPY_TYPES;
+
+/// A newtype wrapper around [`[u8; N]`][Py_UCS1] to handle [`byte` scalars][numpy-bytes] while satisfying coherence.
+///
+/// Note that when creating arrays of ASCII strings without an explicit `dtype`, e.g.
+///
+/// ```python
+/// numpy.array([b"foo", b"bar", b"foobar"])
+/// ```
+///
+/// yielding `S6`, NumPy will automatically determine the smallest possible array length at runtime.
+/// This does not work well with the length `N` of `PyFixedString<N>` determined at compile-time
```

Comment:
A bit minor comment, but `This does not work well with the length `N`` seems ambiguous, because this should actually work with `N=6`. What we need to warn users is that the numpy array should have the same dtype in Rust code, so explicitly giving dtype is better.

Reply:
I tried to rewrite this coming at it from the other end. Please have another look if this is clearer now.

Reply:
Thank you for the rewrite, it looks clear now.